### PR TITLE
Add ALL_PROXY to build-arg whitelist

### LINKED ISSUE #21319: Feature reuqest: Add ALL_PROXY to list of supported variables for Build-time arguments
In https://github.com/docker/docker/issues/14634, variables were whitelisted to add support for proxying via cURL and other libraries. `ALL_PROXY` should be added to that list. A possible bonus feature is to emulate ALL by adding entries for HTTP, HTTPS, and FTP.

### PR DESCRIPTION
**Pull request by thaJeztah**
fixes https://github.com/moby/moby/issues/21319

### CODE REVIEW

**Comment by thaJeztah** [comment]
@tonistiigi @AkihiroSuda FYI (if we're good with this, perhaps changes are needed in BuildKit as well?)

File: builder/dockerfile/buildargs.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ var builtinAllowedBuildArgs = map[string]bool{
 	"https_proxy": true,
 	"FTP_PROXY":   true,
 	"ftp_proxy":   true,
+	"ALL_PROXY":   true,
+	"all_proxy":   true,
```
Comment (AkihiroSuda) [inline builder/dockerfile/buildargs.go]: And SOCKS_PROXY and RSYNC_PROXY?

**Comment by tonistiigi** [comment]
Is there some documentation for these or some authority recommending a list that should be supported (and that we could link to from code and docs). If I search for this, different projects have different sets/recommendations. If we just keep adding things to that list, users will eventually start hitting these accidentally and get failing builds.

File: builder/dockerfile/buildargs.go (hunk 1 above)
Comment (thaJeztah) [reply builder/dockerfile/buildargs.go]: Didn't know about those.. hmf.. not happy there's so many

**Comment by thaJeztah** [comment]
> Is there some documentation for these or some authority recommending a list that should be supported (and that we could link to from code and docs)

Unfortunately, I don't think there is (IIRC, we looked into that when reviewing the `--build-arg` pull request). Yes, I agree it's "messy" (especially given that not even the correct _case_ (upper/lower) seems to be agreed on (or defined [which one takes precedence](https://github.com/curl/curl/issues/2477)).

`HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` are quite common (and even there, [the format for `NO_PROXY` is non-standardised](https://github.com/curl/curl/issues/1208))

Golang (out of the box) seems to only use those; https://github.com/golang/go/blob/8e9386db0ee72ffa847b1673d15f6092291724d9/src/net/http/transport.go#L258-L308, and [require custom diapers for (say) Socks proxy](https://github.com/golang/go/commit/da32ed7bf18919f5397e620b794afe30a7c1dd37)


>  If we just keep adding things to that list, users will eventually start hitting these accidentally and get failing builds.

Not sure I understand 😅 could you elaborate a bit? If they configure these through `ENV`, nothing would change, only when setting then as `ARG` / `--build-arg`


Having said that: yes, I agree it's tricky finding the right balance as to which of these are "common" / "de-facto" variables, and which ones not.


@tianon @AkihiroSuda any suggestions?

**Comment by tonistiigi** [comment]
> Not sure I understand 😅 could you elaborate a bit?

Yes, I meant that these pick up different caching rules but I guess it doesn't matter much because if they redefine it as `ARG` they will start behaving like regular arguments. So that only leaves the case when an unexpected environment variable is exposed to `RUN` what probably isn't very important until they have `_PROXY` in the name.

**Comment by thaJeztah** [comment]
Discussing this with other maintainers, and we didn't get more requests and actual use-cases for this; users can already add them as a custom `ARG`, so there's no big reason to add this. Let's close for now

**Pull request closed without merging**